### PR TITLE
Added train_with_arcface with the new training implementation

### PR DESCRIPTION
**Pull request by hareemraza**
Modified the arcface_loss using Maria's modified pipeline, new iresnet implementation and Alexander's feedback. There is a very slight improvement in the way arcface_loss behaves but the accuracy shows no improvement so far. 

Open to any feedback.

### CODE REVIEW

File: src/train_with_arcface.py
Comment (Matematikoi) [inline src/train_with_arcface.py:1]: This shouldn't be a new file. Add a new parameter to the script that allows us to change between losses as we change between size of epochs.

File: src/train_with_arcface.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        x = self.base_model.bn2(x)
+
+        x = self.base_model.avgpool(x)  # Output shape (batch_size, 512, 1, 1)
+        x = torch.flatten(x, 1)  # Flatten to shape (batch_size, 512)
+        x = self.base_model.features(x)
+        return x
+
+class ArcFace(torch.nn.Module):
+    def __init__(self, s=64.0, margin=0.5):
+        super(ArcFace, self).__init__()
```
Comment (Matematikoi) [inline src/train_with_arcface.py:38]: are these parameters something we should worry about in the future? or we are probably never changing them?

**Review by Matematikoi — CHANGES_REQUESTED** [review]
Its hard to review since it is a completely new file. We should probably just have a parameter to switch between losses since everything else is the same and we are not planning on maintaining two codes at the same time

File: src/train_with_arcface.py
Comment (hareemraza) [reply src/train_with_arcface.py:1]: Done!
Now we have a single train file with an option to choose between losses

File: src/train_with_arcface.py (hunk 1 above)
Comment (hareemraza) [reply src/train_with_arcface.py]: We dont need to change them. Experimented with alot of parameters. These are the defaults ones used by insightface itself (also recommended by Alexander)

File: src/train.py [hunk 2]
```diff
@@ -1,21 +1,75 @@
...
 from pathlib import Path
 from iresnet import iresnet18
 
+# RenNet Model Definition
+class IResNetModified(nn.Module):
+    def __init__(self):
+        super(IResNetModified, self).__init__()
+        self.base_model = iresnet18(pretrained=False, progress=True) 
+        # Remove the fully connected layer for a 512-dimensional embedding
+        self.base_model.fc = nn.Identity()
```
Comment (Matematikoi) [inline src/train.py:19]: you are not using this layer

**Review by Matematikoi — COMMENTED** [review]
just a small issue with the resnet architecture

File: src/train.py [hunk 3]
```diff
@@ -44,67 +98,74 @@ def load_data(train_folder, test_folder, batch_size, num_workers):
...
+            loss = loss_fn(logits, y)
+
+        # Backward and optimize
         loss.backward()
-        optimizer.step()
+        
+        # Just to try gradient clipping, if needed
+        # if use_arcface:
+        #     torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0) 
+        # optimizer.step()
```
Comment (afm215) [inline src/train.py:155]: This should be uncommented, if not weights are never updated

File: src/train.py [hunk 4]
```diff
@@ -133,52 +201,59 @@ def make_checkpoint_dir():
...
+        print(f"\n--- Epoch {epoch}/{num_epochs} ---")
+        avg_train_loss, avg_train_accuracy = train_one_epoch(epoch, model, train_loader, optimizer, loss_fn, device, use_arcface)
         train_loss.append(avg_train_loss)
         train_accuracy.append(avg_train_accuracy)
 
-        if num_epoch % checkpoint_interval == 0:
-            checkpoint_path = f'{datestring}/checkpoint_epoch_{num_epoch}.pt'
+        # Step the scheduler only if it is used
+        if scheduler:
+            scheduler.step()
```
Comment (afm215) [inline src/train.py:223]: This should be put into the epoch loop over batches

File: src/train.py (hunk 3 above)
Comment (hareemraza) [reply src/train.py]: Just cross-checked. This was commented by mistake (while commenting gradient clipping). Fixed in the latest commit.
